clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java

```diff
@@ -158,6 +166,147 @@ public void testCloseAll() {
+            // directly assign the last mocked bytes read to fix `bufferSize` exactly
+            final int mockedBytesRead = (diff < step) ? diff : random.nextInt(step);
+            EasyMock.expect(channelMock.read(EasyMock.anyObject(ByteBuffer.class), EasyMock.anyInt())).andAnswer(new IAnswer<Integer>() {
+                @Override
+                public Integer answer() throws Throwable {
+                    ByteBuffer buffer = (ByteBuffer) EasyMock.getCurrentArguments()[0];
+                    Long pos = (Long) EasyMock.getCurrentArguments()[1];
+                    buffer.position(pos.intValue() + mockedBytesRead);
+                    return mockedBytesRead;
+                }
+            });
+            totalMockedBytesRead += mockedBytesRead;
+        }
+
+        EasyMock.replay(channelMock);
+        Utils.readFullyOrFail(channelMock, buffer, 0L);
+        Assert.assertTrue("The buffer should be filled up", !buffer.hasRemaining());
+        EasyMock.verify(channelMock);
+    }
+
+    @Test
+    public void testReadFullyOrFailThrowIOExceptionForEOF() throws IOException {
+        FileChannel channelMock = EasyMock.createMock(FileChannel.class);
+        ByteBuffer buffer = ByteBuffer.allocate(12);
+        EasyMock.expect(channelMock.read(buffer, 0L)).andReturn(-1);
+        EasyMock.replay(channelMock);
+
+        try {
+            Utils.readFullyOrFail(channelMock, buffer, 0L);
+            fail("Expected IOException to be raised");
+        } catch (IOException e) {
+            // expected
+        }
+        EasyMock.verify(channelMock);
+    }
+
+    @Test
+    public void testReadFullyWithMultiReads() throws IOException {
+        FileChannel channelMock = EasyMock.createMock(FileChannel.class);
+        final int bufferSize = 100;
+        final int step = 20;
+        ByteBuffer buffer = ByteBuffer.allocate(bufferSize);
+        final Random random = new Random();
+        int totalMockedBytesRead = 0;
+
+        while (true) {
+            int diff = bufferSize - totalMockedBytesRead;
+            if (diff <= 0) {
+                break;
+            }
+            // If the remaining byte number is less than a step,
+            // directly assign the last mocked bytes read to fix `bufferSize` exactly
+            final int mockedBytesRead = (diff < step) ? diff : random.nextInt(step);
+            EasyMock.expect(channelMock.read(EasyMock.anyObject(ByteBuffer.class), EasyMock.anyInt())).andAnswer(new IAnswer<Integer>() {
+                @Override
+                public Integer answer() throws Throwable {
+                    ByteBuffer buffer = (ByteBuffer) EasyMock.getCurrentArguments()[0];
+                    Long pos = (Long) EasyMock.getCurrentArguments()[1];
+                    buffer.position(pos.intValue() + mockedBytesRead);
+                    return mockedBytesRead;
+                }
+            });
+            totalMockedBytesRead += mockedBytesRead;
+        }
```

Comment:
Seems like this is the same in `testReadFullyOrFailWithMultiReads`. Maybe we can extract it to a helper method.